pkg/model/multiversion/delta.go

```diff
@@ -0,0 +1,248 @@
+	if err != nil {
+		return nil, fmt.Errorf("cannot compute status fields deltas: %s", err)
+	}
+
+	return &CRDDelta{
+		specDeltas,
+		statusDeltas,
+	}, nil
+}
+
+func max(a, b int) int {
+	if a > b {
+		return a
+	}
+	return b
+}
+
+// ComputeFieldsDiff computes the difference between two maps of fields. It returns a list
+// of FieldDelta's that contains the the ChangeType and at least one field reference.
+func ComputeFieldsDiff(
+	spokeCRDFields map[string]*ackmodel.Field,
+	hubCRDFields map[string]*ackmodel.Field,
+	// the hub renames
+	renames map[string]string,
+) ([]FieldDelta, error) {
+	// the smallest size of deltas we can get is max(len(spokeCRDFields), len(hubCRDFields))
+	deltas := make([]FieldDelta, 0, max(len(spokeCRDFields), len(hubCRDFields)))
+
+	// collect field names and sort them to ensure a determenistic output order.
+	spokeFieldsNames := []string{}
+	for name := range spokeCRDFields {
+		spokeFieldsNames = append(spokeFieldsNames, name)
+	}
+	sort.Strings(spokeFieldsNames)
+
+	hubFieldsNames := []string{}
+	for name := range hubCRDFields {
+		hubFieldsNames = append(hubFieldsNames, name)
+	}
+	sort.Strings(hubFieldsNames)
+
+	// let's make sure we don't visit fields more than once - especially
+	// when fields are renamed.
+	visitedFields := map[string]*struct{}{}
+
+	// first let's loop over the spokeFieldsNames array and see if we can find
+	// the same field name in hubFieldNames.
+	for _, spokeFieldName := range spokeFieldsNames {
+		spokeField, _ := spokeCRDFields[spokeFieldName]
+		hubField, ok := hubCRDFields[spokeFieldName]
+		// If a field is found in both arrays only three changes are possible:
+		// Intact, TypeChangeChange and ChangeTypeShapeChangedToSecret.
+		// NOTE(a-hilaly): carefull about X -> Y then Z -> X renames. It should
+		// not be allowed.
+		if ok {
+			// mark field as visited.
+			visitedFields[spokeFieldName] = nil
+			// check if field became secret
+			if (spokeField.FieldConfig == nil ||
+				(spokeField.FieldConfig != nil && !spokeField.FieldConfig.IsSecret)) &&
+				(hubField.FieldConfig != nil && hubField.FieldConfig.IsSecret) {
+				deltas = append(deltas, FieldDelta{
+					Spoke:      spokeField,
+					Hub:        hubField,
+					ChangeType: FieldChangeTypeShapeChangedToSecret,
+				})
+				continue
+			}
+
+			// TODO(a-hilaly) better function for equalizing shapes
+			equalShapes, err := isEqualShape(spokeField.ShapeRef, hubField.ShapeRef)
+			if err != nil {
+				return nil, err
+			}
+			if equalShapes {
+				// if the fields have equal names and types the change is intact
+				deltas = append(deltas, FieldDelta{
+					Spoke:      spokeField,
+					Hub:        hubField,
+					ChangeType: FieldChangeTypeIntact,
+				})
+				continue
+			}
+
+			// at this point we know that the fields kept the same name but have different
+			// shapes
+			deltas = append(deltas, FieldDelta{
+				Spoke:      spokeField,
+				Hub:        hubField,
+				ChangeType: FieldChangeTypeIntact,
```

Comment:
copy paste error? This should be "FieldChangeTypeShapeChanged". correct?

Reply:
Yup nice catch!